### PR TITLE
[Distributed] correct take semantics for synthesized ID assignments

### LINKED ISSUE #62898: "SIL memory lifetime failure" when attempting to build Distributed Actors example.
**Description**
When attempting to run an example from the Distributed Actors repository (https://github.com/apple/swift-distributed-actors/), the compilation process encounters an error:

```
SIL memory lifetime failure in @$s18DistributedCluster0B16EventStreamActorC11actorSystemAcA0bG0C_tcfc: memory is initialized, but shouldn't be
memory location:   %40 = alloc_stack $ClusterSystem.ActorID        // users: %51, %50, %45
at instruction:   dealloc_stack %40 : $*ClusterSystem.ActorID     // id: %51
```

Further down, the additional error information is reported:
```
1.	Apple Swift version 5.8-dev (LLVM ed4d811c2897d77, Swift 8491f52d478aca6)
2.	Compiling with the current language version
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for module DistributedCluster)
4.	While silgen constructor initializer SIL function "@$s18DistributedCluster0B16EventStreamActorC11actorSystemAcA0bG0C_tcfc".
 for 'init(actorSystem:)' (at /Users/username/distActors/Sources/DistributedCluster/Cluster/ClusterEventStream.swift:152:14)
5.	While verifying SIL function "@$s18DistributedCluster0B16EventStreamActorC11actorSystemAcA0bG0C_tcfc".
 for 'init(actorSystem:)' (at /Users/username/distActors/Sources/DistributedCluster/Cluster/ClusterEventStream.swift:152:14)
```

Stack dump:
```
0  swift-frontend           0x0000000105462c3c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x0000000105461ec0 llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x000000010546327c SignalHandler(int) + 304
3  libsystem_platform.dylib 0x000000018d1314c4 _sigtramp + 56
4  libsystem_pthread.dylib  0x000000018d119ee0 pthread_kill + 288
5  libsystem_c.dylib        0x000000018d054340 abort + 168
6  swift-frontend           0x00000001054ccd2c (anonymous namespace)::MemoryLifetimeVerifier::reportError(llvm::Twine const&, int, swift::SILInstruction*) (.cold.2) + 0
7  swift-frontend           0x0000000100d9a1b8 std::__1::__function::__func<(anonymous namespace)::MemoryLifetimeVerifier::verify()::$_0, std::__1::allocator<(anonymous namespace)::MemoryLifetimeVerifier::verify()::$_0>, void (swift::SILBasicBlock*)>::~__func() + 0
8  swift-frontend           0x0000000100d999f4 (anonymous namespace)::MemoryLifetimeVerifier::require(llvm::SmallBitVector const&, llvm::Twine const&, swift::SILInstruction*, bool) + 1444
9  swift-frontend           0x0000000100d98ec4 (anonymous namespace)::MemoryLifetimeVerifier::checkBlock(swift::SILBasicBlock*, llvm::SmallBitVector&) + 3980
10 swift-frontend           0x0000000100d9a3c8 std::__1::__function::__func<(anonymous namespace)::MemoryLifetimeVerifier::verify()::$_0, std::__1::allocator<(anonymous namespace)::MemoryLifetimeVerifier::verify()::$_0>, void (swift::SILBasicBlock*)>::operator()(swift::SILBasicBlock*&&) + 436
11 swift-frontend           0x0000000100d6c634 swift::MemoryLocations::handleSingleBlockLocations(std::__1::function<void (swift::SILBasicBlock*)>) + 180
12 swift-frontend           0x0000000100d97a2c swift::SILFunction::verifyMemoryLifetime() + 4880
13 swift-frontend           0x0000000100da4818 (anonymous namespace)::SILVerifier::visitSILFunction(swift::SILFunction*) + 8856
14 swift-frontend           0x0000000100d9e74c swift::SILFunction::verify(bool) const + 76
15 swift-frontend           0x000000010115ffdc swift::Lowering::SILGenModule::postEmitFunction(swift::SILDeclRef, swift::SILFunction*) + 228
16 swift-frontend           0x000000010115f438 swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 6928
17 swift-frontend           0x0000000101160ef8 emitOrDelayFunction(swift::Lowering::SILGenModule&, swift::SILDeclRef, bool) + 384
18 swift-frontend           0x0000000101161110 swift::Lowering::SILGenModule::emitConstructor(swift::ConstructorDecl*) + 352
19 swift-frontend           0x000000010121d4bc (anonymous namespace)::SILGenType::emitType() + 400
20 swift-frontend           0x000000010121d320 swift::Lowering::SILGenModule::visitNominalTypeDecl(swift::NominalTypeDecl*) + 24
21 swift-frontend           0x0000000101163068 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const + 1788
22 swift-frontend           0x0000000101212e08 swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&) + 148
23 swift-frontend           0x0000000101166f28 llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&) + 400
24 swift-frontend           0x0000000101163b70 swift::performASTLowering(swift::ModuleDecl*, swift::Lowering::TypeConverter&, swift::SILOptions const&, swift::IRGenOptions const*) + 140
25 swift-frontend           0x0000000100bdee94 swift::performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 1060
26 swift-frontend           0x0000000100beda5c withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) + 160
27 swift-frontend           0x0000000100be09d4 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3256
28 swift-frontend           0x0000000100a49828 swift::mainEntry(int, char const**) + 3136
29 dyld                     0x000000010e9a1088 start + 516
```

**Steps to reproduce**
1. Create a new directory.
2. Inside the new directory, run: `git clone https://github.com/apple/swift-distributed-actors.git`
3. Run: `swift run --package-path Samples/Sources/SampleDiningPhilosophers`

The commit number for the distributed actors repository in this case is `247f23a`.

**Expected behavior**
The compilation process should succeed, and the example should run.

**Environment**
This happens to fail with toolchains: 2022-09-27a, 2022-11-03a, and 2023-01-02a on an M1. Also fails with 2022-11-03a on x86.

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/apple/swift/issues/62898
Resolves rdar://104583893

This is a surprising regression because it always was wrong, but only started failing in recent toolchains where this started getting verified.

I'll also wok on getting distributed cluster into the source compat build, which we didn't complete for some reason before: https://github.com/apple/swift-source-compat-suite/pull/730 which would have likely caught this when the toolchain changed to detect this wrong lifetime.

I also confirmed this works on the reproducer steps, targeting swift-distributed actors `main` and indeed this started failing but used to somehow sneakily still pass on older toolchains:
- `git clone https://github.com/apple/swift-distributed-actors.git`
- `export SWIFT_EXEC=$HOME/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swiftc`
    - on a build that has this fix
- `swift run --package-path Samples/Sources/SampleDiningPhilosophers`
- `echo $SWIFT_EXEC; echo; echo; /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2023-02-18-a.xctoolchain/usr/bin/swift run --package-path Samples/Sources/SampleDiningPhilosophers`

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: test/Distributed/SIL/distributed_actor_default_init_sil_8.swift [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  init() throws {
+    do {
+      actorSystem = try getSystem()
+    } catch {
+      actorSystem = FakeRoundtripActorSystem()
+    }
+    someField = try getSomeClass()
+  }
+
```
Comment (ktoso) [inline test/Distributed/SIL/distributed_actor_default_init_sil_8.swift:37]: This adds a test for a class based actor system which was a thing we didn't cover well, but in the real world most systems will be; This is the same as the other sil_n tests but for a class.

File: lib/SILGen/SILGenDistributed.cpp [hunk 2]
```diff
@@ -209,7 +209,7 @@ void SILGenFunction::emitDistActorIdentityInit(ConstructorDecl *ctor,
       { temp, selfMetatypeValue });
 
   // --- initialize the property.
-  initializeProperty(*this, loc, borrowedSelfArg, var, temp);
```
Comment (ktoso) [inline lib/SILGen/SILGenDistributed.cpp:212]: The issue was here, so changing to take here is correct, but the other sites want to remain not-take.

Thanks @eeckstein and @drexin for the pointers !

**Review by drexin — APPROVED** [review]
LGTM

**Review by xedin — APPROVED** [review]

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: test/Distributed/Inputs/FakeDistributedActorSystems.swift [hunk 3]
```diff
@@ -34,6 +34,36 @@ public struct ActorAddress: Hashable, Sendable, Codable {
   }
 }
 
+/// This type is same as ActorAddress however for purposes of SIL tests we make it not-loadable.
+///
+/// By adding the `Any` we don't know the full size of the struct making the type in SIL `$*ActorAddress`
+/// when we try to store or pass it around, which triggers `isAddressOnly` guarded paths which we need to test.
+public struct NotLoadableActorAddress: Hashable, Sendable, Codable {
```
Comment (ktoso) [inline test/Distributed/Inputs/FakeDistributedActorSystems.swift:41]: The important bit to reproduce the issue

**Comment by ktoso** [comment]
Added another test that exercises the AddressOnly case as well.

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Pull request merged**